<?xml version="1.0" ?>
<!--
  This is not a standalone urdf for kobuki. It simply defines the 'kobuki' tag that can
  be incorporated by other urdf files (e.g. turtlebot).

  See kobuki_standalone.urdf.xacro for a standalone urdf to be tested with
      kobuki_description/launch/view_model.launch
 -->
<robot name="kobuki" xmlns:xacro="http://ros.org/wiki/xacro">
  <xacro:include filename="$(find kobuki_description)/urdf/common_properties.urdf.xacro"/>
  <xacro:include filename="$(find kobuki_description)/urdf/kobuki_gazebo.urdf.xacro"/>

  <!-- Kobuki -->
  <xacro:macro name="kobuki">
    <link name="base_footprint"/>
    <!--
       Base link is set at the bottom of the base mould.
       This is done to be compatible with the way base link
       was configured for turtlebot 1. Refer to

       https://github.com/turtlebot/turtlebot/issues/40

       To put the base link at the more oft used wheel
       axis, set the z-distance from the base_footprint
       to 0.352.
      -->
    <joint name="base_joint" type="fixed">
      <origin xyz="0 0 0.0102" rpy="0 0 0" />
      <parent link="base_footprint"/>
      <child link="base_link" />
    </joint>
    <link name="base_link">
      <visual>
        <geometry>
          <!-- new mesh -->
          <mesh filename="package://kobuki_description/meshes/main_body.dae" />
        </geometry>
        <origin xyz="0.001 0 0.05199" rpy="0 0 0"/>
      </visual>
      <collision name="base">
        <geometry>
          <cylinder length="0.10938" radius="0.178"/>
        </geometry>
        <origin xyz="0.0 0 0.05949" rpy="0 0 0"/>
      </collision>
      <inertial>
        <!-- COM experimentally determined -->
        <origin xyz="0.01 0 0"/>
        <mass value="2.4"/> <!-- 2.4/2.6 kg for small/big battery pack -->
        <!-- Kobuki's inertia tensor is approximated by a cylinder with homogeneous mass distribution
             More details: http://en.wikipedia.org/wiki/List_of_moment_of_inertia_tensors
             m = 2.4 kg; h = 0.09 m; r = 0.175 m
             ixx = 1/12 * m * (3 * r^2 + h^2)
             iyy = 1/12 * m * (3 * r^2 + h^2)
             izz = 1/2 * m * r^2
          -->
        <inertia ixx="0.019995" ixy="0.0" ixz="0.0"
                 iyy="0.019995" iyz="0.0"
                 izz="0.03675" />
      </inertial>
    </link>

    <joint name="wheel_left_joint" type="continuous">
      <parent link="base_link"/>
      <child link="wheel_left_link"/>
      <origin xyz="0.00 ${0.23/2} 0.0250" rpy="${-M_PI/2} 0 0"/>
      <axis xyz="0 0 1"/>
    </joint>
    <link name="wheel_left_link">
      <visual>
        <geometry>
          <mesh filename="package://kobuki_description/meshes/wheel.dae"/>
        </geometry>
        <origin xyz="0 0 0" rpy="0 0 0"/>
      </visual>
      <collision>
        <geometry>
          <cylinder length="0.0206" radius="0.035"/>
        </geometry>
        <origin rpy="0 0 0" xyz="0 0 0"/>
      </collision>
      <inertial>
        <mass value="0.01" />
        <origin xyz="0 0 0" />
        <inertia ixx="0.001" ixy="0.0" ixz="0.0"
                 iyy="0.001" iyz="0.0"
                 izz="0.001" />
      </inertial>
    </link>

    <joint name="wheel_right_joint" type="continuous">
      <parent link="base_link"/>
      <child link="wheel_right_link"/>
      <origin xyz="0.00 -${0.23/2} 0.0250" rpy="${-M_PI/2} 0 0"/>
      <axis xyz="0 0 1"/>
    </joint>
    <link name="wheel_right_link">
      <visual>
        <geometry>
          <mesh filename="package://kobuki_description/meshes/wheel.dae"/>
        </geometry>
        <origin xyz="0 0 0" rpy="0 0 0"/>
      </visual>
      <collision>
        <geometry>
          <cylinder length="0.0206" radius="0.035"/>
        </geometry>
        <origin rpy="0 0 0" xyz="0 0 0"/>
      </collision>
      <inertial>
        <mass value="0.01" />
        <origin xyz="0 0 0" />
        <inertia ixx="0.001" ixy="0.0" ixz="0.0"
                 iyy="0.001" iyz="0.0"
                 izz="0.001" />
      </inertial>
    </link>

    <joint name="caster_front_joint" type="fixed">
      <parent link="base_link"/>
      <child link="caster_front_link"/>
      <origin xyz="0.115 0.0 0.007" rpy="${-M_PI/2} 0 0"/>
    </joint>
    <link name="caster_front_link">
      <collision>
        <geometry>
          <cylinder length="0.0176" radius="0.017"/>
        </geometry>
        <origin rpy="0 0 0" xyz="0 0 0"/>
      </collision>
      <inertial>
        <mass value="0.01" />
        <origin xyz="0 0 0" />
        <inertia ixx="0.001" ixy="0.0" ixz="0.0"
                 iyy="0.001" iyz="0.0"
                 izz="0.001" />
      </inertial>
    </link>

    <joint name="caster_back_joint" type="fixed">
      <parent link="base_link"/>
      <child link="caster_back_link"/>
      <origin xyz="-0.135 0.0 0.009" rpy="${-M_PI/2} 0 0"/>
    </joint>
    <link name="caster_back_link">
      <collision>
        <geometry>
          <cylinder length="0.0176" radius="0.017"/>
        </geometry>
        <origin rpy="0 0 0" xyz="0 0 0"/>
      </collision>
      <inertial>
        <mass value="0.01" />
        <origin xyz="0 0 0" />
        <inertia ixx="0.001" ixy="0.0" ixz="0.0"
                 iyy="0.001" iyz="0.0"
                 izz="0.001" />
      </inertial>
    </link>

    <!-- Kobuki's sensors -->
    <joint name="gyro_joint" type="fixed">
      <axis xyz="0 1 0"/>
      <origin xyz="0.056 0.062 0.0202" rpy="0 0 0"/>
      <parent link="base_link"/>
      <child link="gyro_link"/>
    </joint>
    <link name="gyro_link">
      <inertial>
        <mass value="0.001"/>
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <inertia ixx="0.0001" ixy="0" ixz="0"
                 iyy="0.000001" iyz="0"
                 izz="0.0001"/>
      </inertial>
    </link>

    <joint name="cliff_sensor_left_joint" type="fixed">
      <origin xyz="0.08734 0.13601 0.0214" rpy="0 ${M_PI/2} 0" />
      <parent link="base_link"/>
      <child link="cliff_sensor_left_link" />
    </joint>
    <link name="cliff_sensor_left_link">
      <inertial>
        <mass value="0.0001" />
        <origin xyz="0 0 0" />
        <inertia ixx="0.0001" ixy="0.0" ixz="0.0"
                 iyy="0.0001" iyz="0.0"
                 izz="0.0001" />
      </inertial>
    </link>

    <joint name="cliff_sensor_right_joint" type="fixed">
      <origin xyz="0.085 -0.13601 0.0214" rpy="0 ${M_PI/2} 0" />
      <parent link="base_link"/>
      <child link="cliff_sensor_right_link" />
    </joint>
    <link name="cliff_sensor_right_link">
      <inertial>
        <mass value="0.0001" />
        <origin xyz="0 0 0" />
        <inertia ixx="0.0001" ixy="0.0" ixz="0.0"
                 iyy="0.0001" iyz="0.0"
                 izz="0.0001" />
      </inertial>
    </link>

    <joint name="cliff_sensor_front_joint" type="fixed">
      <origin xyz="0.156 0.00 0.0214" rpy="0 ${M_PI/2} 0" />
      <parent link="base_link"/>
      <child link="cliff_sensor_front_link" />
    </joint>
    <link name="cliff_sensor_front_link">
      <inertial>
        <mass value="0.0001" />
        <origin xyz="0 0 0" />
        <inertia ixx="0.0001" ixy="0.0" ixz="0.0"
                 iyy="0.0001" iyz="0.0"
                 izz="0.0001" />
      </inertial>
    </link>

    <!-- FEATURED ROBOT MODEL -->
    <!-- head pan_tilt -->
    <!-- head base -->
    <link name="head_base_link">
      <visual>
        <geometry>
          <box size="0.045 0.03 0.035"/>
        </geometry>
        <origin xyz="0.0 0.0 0.0" rpy="0 0 0" />
        <material name="blue">
            <color rgba="0 0 0.8 1"/>
        </material>
      </visual>
    </link>

    <joint name="head_base_joint" type="fixed">
      <origin xyz="-0.20 0.00 0.90" rpy="0 0 0" />
      <parent link="base_link"/>
      <child link="head_base_link" />
    </joint>

    <!-- head pan -->
    <link name="head_pan_link">
      <visual>
        <geometry>
          <box size="0.03 0.035 0.045"/>
        </geometry>
        <origin xyz="0.0 0.0 0.0" rpy="0 0 0" />
        <material name="black">
            <color rgba="0 0 0 1"/>
        </material>
      </visual>
    </link>

    <joint name="head_pan_joint" type="revolute">
      <axis xyz="0 0 1"/>
      <limit effort="1000.0" lower="-3.1416" upper="3.1416" velocity="0.5"/>
      <origin xyz="0.0 0.0 0.040" rpy="0 0 0" />
      <parent link="head_base_link"/>
      <child link="head_pan_link" />
    </joint>

    <!-- head tilt -->
    <link name="head_tilt_link">
      <visual>
        <geometry>
          <box size="0.025 0.045 0.025"/>
        </geometry>
        <origin xyz="0.0 0.0 0.0" rpy="0 0 0" />
        <material name="white">
            <color rgba="1 1 1 1"/>
        </material>
      </visual>
    </link>

    <joint name="head_tilt_joint" type="revolute">
      <axis xyz="0 1 0"/>
      <limit effort="1000.0" lower="-3.1416" upper="3.1416" velocity="0.5"/>
      <origin xyz="0.00 0.00 0.035" rpy="0 0 0" />
      <parent link="head_pan_link"/>
      <child link="head_tilt_link" />
    </joint>

    <!-- camera -->
    <link name="camera_rgb_frame"/>
    <joint name="camera_rgb_joint" type="fixed">
      <origin xyz="0.0 0.0 0.05" rpy="0 0 0" />
      <parent link="head_tilt_link"/>
      <child link="camera_rgb_frame" />
    </joint>

    <!-- robot arm as in built as in http://wiki.ros.org/turtlebot_arm -->
    <!-- arm base -->
    <link name="arm_base_link">
      <visual>
        <geometry>
          <box size="0.045 0.03 0.035"/>
        </geometry>
        <origin xyz="0.0 0.0 0.0" rpy="0 0 0" />
        <material name="blue">
            <color rgba="0 0 0.8 1"/>
        </material>
      </visual>
    </link>

    <joint name="arm_base_joint" type="fixed">
      <origin xyz="0.10 0.00 0.43" rpy="0 0 0" />
      <parent link="base_link"/>
      <child link="arm_base_link" />
    </joint>

    <!-- arm waist -->
    <link name="arm_waist_link">
      <visual>
        <geometry>
          <box size="0.03 0.035 0.045"/>
        </geometry>
        <origin xyz="0.0 0.0 0.0" rpy="0 0 0" />
        <material name="black">
            <color rgba="0 0 0 1"/>
        </material>
      </visual>
    </link>

    <joint name="arm_waist_joint" type="revolute">
      <axis xyz="0 0 1"/>
      <limit effort="1000.0" lower="-3.1416" upper="3.1416" velocity="0.5"/>
      <origin xyz="0.0 0.0 0.040" rpy="0 0 0" />
      <parent link="arm_base_link"/>
      <child link="arm_waist_link" />
    </joint>

    <!-- arm shoulder -->
    <link name="arm_shoulder_link">
      <visual>
        <geometry>
          <box size="0.025 0.045 0.05"/>
        </geometry>
        <origin xyz="0.0 0.0 0.0" rpy="0 0 0" />
        <material name="white">
            <color rgba="1 1 1 1"/>
        </material>
      </visual>
    </link>

    <joint name="arm_shoulder_joint" type="revolute">
      <axis xyz="0 1 0"/>
      <limit effort="1000.0" lower="-3.1416" upper="3.1416" velocity="0.5"/>
      <origin xyz="0.00 0.00 0.0475" rpy="0 0 0" />
      <parent link="arm_waist_link"/>
      <child link="arm_shoulder_link" />
    </joint>

    <!-- arm elbow -->
    <link name="arm_elbow_base_link">
      <visual>
        <geometry>
          <box size="0.03 0.035 0.045"/>
        </geometry>
        <origin xyz="0.0 0.0 0.0" rpy="0 0 0" />
        <material name="black">
            <color rgba="0 0 0 1"/>
        </material>
      </visual>
    </link>

    <joint name="arm_elbow_base_joint" type="fixed">
      <origin xyz="0.0 0.00 0.0475" rpy="0 0 0" />
      <parent link="arm_shoulder_link"/>
      <child link="arm_elbow_base_link" />
    </joint>

    <link name="arm_elbow_link">
      <visual>
        <geometry>
          <box size="0.025 0.045 0.05"/>
        </geometry>
        <origin xyz="0.0 0.0 0.0" rpy="0 0 0" />
        <material name="white">
            <color rgba="1 1 1 1"/>
        </material>
      </visual>
    </link>

    <joint name="arm_elbow_joint" type="revolute">
      <axis xyz="0 1 0"/>
      <limit effort="1000.0" lower="-3.1416" upper="3.1416" velocity="0.5"/>
      <origin xyz="0.00 0.00 0.0475" rpy="0 0 0" />
      <parent link="arm_elbow_base_link"/>
      <child link="arm_elbow_link" />
    </joint>

    <!-- arm wrist -->
    <link name="arm_wrist_base_link">
      <visual>
        <geometry>
          <box size="0.03 0.035 0.045"/>
        </geometry>
        <origin xyz="0.0 0.0 0.0" rpy="0 0 0" />
        <material name="black">
            <color rgba="0 0 0 1"/>
        </material>
      </visual>
    </link>

    <joint name="arm_wrist_base_joint" type="fixed">
      <origin xyz="0.0 0.00 0.0475" rpy="0 0 0" />
      <parent link="arm_elbow_link"/>
      <child link="arm_wrist_base_link" />
    </joint>

    <link name="arm_wrist_link">
      <visual>
        <geometry>
          <box size="0.025 0.045 0.025"/>
        </geometry>
        <origin xyz="0.0 0.0 0.0" rpy="0 0 0" />
        <material name="white">
            <color rgba="1 1 1 1"/>
        </material>
      </visual>
    </link>

    <joint name="arm_wrist_joint" type="revolute">
      <axis xyz="0 1 0"/>
      <limit effort="1000.0" lower="-3.1416" upper="3.1416" velocity="0.5"/>
      <origin xyz="0.00 0.00 0.035" rpy="0 0 0" />
      <parent link="arm_wrist_base_link"/>
      <child link="arm_wrist_link" />
    </joint>

    <!-- arm hand -->
    <link name="arm_hand_base_link">
      <visual>
        <geometry>
          <box size="0.035 0.045 0.03"/>
        </geometry>
        <origin xyz="0.0 0.0 0.0" rpy="0 0 0" />
        <material name="black">
            <color rgba="0 0 0 1"/>
        </material>
      </visual>
    </link>

    <joint name="arm_hand_base_joint" type="fixed">
      <origin xyz="0.0 0.00 0.0275" rpy="0 0 0" />
      <parent link="arm_wrist_link"/>
      <child link="arm_hand_base_link" />
    </joint>

    <link name="arm_hand_link">
      <visual>
        <geometry>
          <box size="0.005 0.05 0.03"/>
        </geometry>
        <origin xyz="0.0 0.0 0.0" rpy="0 1.57 1.57" />
        <material name="white">
            <color rgba="1 1 1 1"/>
        </material>
      </visual>
    </link>

    <joint name="arm_hand_joint" type="revolute">
      <axis xyz="0 1 0"/>
      <limit effort="1000.0" lower="-3.1416" upper="3.1416" velocity="0.5"/>
      <origin xyz="0.00 -0.0225 0.04" rpy="0 1.57 1.57" />
      <parent link="arm_hand_base_link"/>
      <child link="arm_hand_link" />
    </joint>

    <link name="arm_hand_palm_link">
      <visual>
        <geometry>
          <box size="0.005 0.04 0.03"/>
        </geometry>
        <origin xyz="0.0 0.0 0.0" rpy="0 1.57 1.57" />
        <material name="white">
            <color rgba="1 1 1 1"/>
        </material>
      </visual>
    </link>

    <joint name="arm_hand_palm_joint" type="fixed">
      <origin xyz="0.00 0.0225 0.035" rpy="0 1.57 1.57" />
      <parent link="arm_hand_base_link"/>
      <child link="arm_hand_palm_link" />
    </joint>

    <!-- Kobuki Gazebo simulation details -->
    <kobuki_sim/>

  </xacro:macro>
</robot>
